# D1 - Environment Setup

## Objective
The objective of D1 is to set up a reproducible development environment for the Task Management API project, establish the core repository structure, configure dependencies, and verify the FastAPI and ASGI application entry points using Pydantic settings.

## Concepts Covered
- **Python Environment & Runtime Verification**: Verifying the Python version and workspace paths.
- **FastAPI Framework & ASGI Integration**: Initializing the core application instance and middleware.
- **Pydantic Settings & Environment Variables**: Loading application configuration safely (`BaseSettings`).
- **Modular Project Structure**: Organizing code into clean, scalable packages (`app/`, `tests/`, `scripts/`).

## Project Implementation
The application structure is organized into modular packages:
- `app/main.py`: Creates the FastAPI application instance, configures lifespan events, and registers routers.
- `app/config.py`: Defines `Settings` powered by `pydantic_settings` and `PipelineConfig` Pydantic models.
- `app/api/`: Implements API endpoints for tasks, users, and health checks.

In this demonstration, we inspect the Python environment, project layout, configuration, and app initialization using direct imports from the project.

## Demonstration
We perform environment inspection and verify the project modules using Python code cells executed from the repository root.

In [1]:
import sys
from pathlib import Path

# Resolve project root directory regardless of execution context
root_dir = Path.cwd().resolve()
if root_dir.name == "notebooks":
    root_dir = root_dir.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))

print(f"Python Version: {sys.version}")
print(f"Executable: {sys.executable}")
print(f"Project Root Directory: {root_dir}")

Python Version: 3.13.7 (tags/v3.13.7:bcee1c3, Aug 14 2025, 14:15:11) [MSC v.1944 64 bit (AMD64)]
Executable: C:\Users\Maha Monisha\AppData\Local\Programs\Python\Python313\python.exe
Project Root Directory: C:\Users\Maha Monisha\OneDrive\Desktop\Triton Internship\task-management-api


In [2]:
import sys
from pathlib import Path

# Resolve project root directory regardless of execution context
root_dir = Path.cwd().resolve()
if root_dir.name == "notebooks":
    root_dir = root_dir.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))

# Verify Project Directory Structure
app_dir = root_dir / "app"

print("Project Layout:")
for item in sorted(app_dir.rglob("*.py")):
    rel_path = item.relative_to(root_dir)
    print(f" - {rel_path}")

Project Layout:
 - app\__init__.py
 - app\config.py
 - app\main.py
 - app\models\__init__.py
 - app\models\task.py
 - app\models\user.py
 - app\schemas\__init__.py
 - app\schemas\task.py
 - app\schemas\user.py
 - app\services\__init__.py
 - app\services\pipeline.py
 - app\services\pipeline_stages.py
 - app\services\task_service.py
 - app\services\user_service.py
 - app\utils\__init__.py
 - app\utils\cache.py
 - app\utils\context_managers.py
 - app\utils\decorators.py
 - app\utils\exceptions.py
 - app\utils\file_iterator.py
 - app\utils\helpers.py
 - app\utils\logging_config.py
 - app\utils\validators.py


In [3]:
import sys
from pathlib import Path

# Resolve project root directory regardless of execution context
root_dir = Path.cwd().resolve()
if root_dir.name == "notebooks":
    root_dir = root_dir.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))

# Import and inspect FastAPI App and Settings
from app.main import app
from app.config import settings

print(f"App Title: {app.title}")
print(f"App Version: {app.version}")
print(f"Project Name: {settings.PROJECT_NAME}")
print(f"Host: {settings.HOST}:{settings.PORT}")
print(f"Debug Mode: {settings.DEBUG}")

App Title: Task Management API
App Version: 1.0.0
Project Name: Task Management API
Host: 127.0.0.1:8000
Debug Mode: True


In [4]:
import sys
from pathlib import Path

# Resolve project root directory regardless of execution context
root_dir = Path.cwd().resolve()
if root_dir.name == "notebooks":
    root_dir = root_dir.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))

# Verify FastAPI Routes
from app.main import app

print("Configured Routes:")
for route in app.routes:
    methods = getattr(route, 'methods', None)
    methods_str = f"[{', '.join(methods)}]" if methods else ""
    print(f" - Path: {route.path:<30} {methods_str:<15} Name: {route.name}")

Configured Routes:
 - Path: /openapi.json                  [HEAD, GET]     Name: openapi
 - Path: /docs                          [HEAD, GET]     Name: swagger_ui_html
 - Path: /docs/oauth2-redirect          [HEAD, GET]     Name: swagger_ui_redirect
 - Path: /redoc                         [HEAD, GET]     Name: redoc_html
 - Path: /                              [GET]           Name: read_root
 - Path: /users                         [POST]          Name: create_user_endpoint
 - Path: /users                         [GET]           Name: get_users_endpoint
 - Path: /users/{user_id}               [GET]           Name: get_user_endpoint
 - Path: /tasks                         [POST]          Name: create_task_endpoint
 - Path: /tasks                         [GET]           Name: get_tasks_endpoint
 - Path: /tasks/{task_id}               [GET]           Name: get_task_endpoint
 - Path: /tasks/{task_id}/complete      [PATCH]         Name: update_task_completion_endpoint


## Actual Output
The code cells above demonstrate:
1. Python 3.13 runtime environment details.
2. Complete directory tree under `app/` confirming modular structure.
3. Successfully instantiated `FastAPI` instance (`app`) and `Settings` instance loaded with defaults.
4. Correctly registered application endpoints (`/`, `/health`, `/tasks`, `/users`).

## Key Observations
- The project relies on `pathlib.Path` for cross-platform path resolution.
- Configuration settings are strongly typed and managed via Pydantic `BaseSettings`.
- FastAPI application initialization is clean and verifies the ASGI setup without requiring external server processes during testing.

## Conclusion
D1 environment setup successfully establishes a robust foundation for building high-performance, maintainable APIs using FastAPI and Pydantic.